src/puya/awst_build/module.py

```diff
@@ -172,6 +192,56 @@ def visit_class_def(self, cdef: mypy.nodes.ClassDef) -> None:
                 location=cdef,
             )
 
+    def _map_scratch_space_reservation(
+        self, expr: mypy.nodes.Expression
+    ) -> ScratchSpaceReservation:
+        source_location = self._location(expr)
+        match expr:
+            case mypy.nodes.IntExpr(value):
+                return ScratchSpaceReservation(
+                    start_slot=value,
+                    stop_slot=value + 1,
+                    source_location=source_location,
+                )
+            case mypy.nodes.CallExpr(
+                callee=mypy.nodes.NameExpr(fullname=constants.URANGE), args=args
+            ):
+                match args:
+                    case [stop]:
+                        stop_int = self._read_int_constant(stop)
+                        return ScratchSpaceReservation(
+                            start_slot=0,
+                            stop_slot=stop_int,
+                            source_location=source_location,
+                        )
+                    case [start, stop]:
+                        start_int = self._read_int_constant(start)
+                        stop_int = self._read_int_constant(stop)
+                        return ScratchSpaceReservation(
+                            start_slot=start_int,
+                            stop_slot=stop_int,
+                            source_location=source_location,
+                        )
+                    case [_, _, _]:
+                        raise CodeError(
+                            "Stride is not supported when specifying ranges of reserved scratch "
+                            "slots"
+                        )
+                    case _:
+                        raise CodeError("Unexpected arguments for urange")
+
+            case _:
+                raise CodeError("Unexpected value: Expected int literal or urange expression")
+
+    def _read_int_constant(self, expr: mypy.nodes.Expression) -> int:
```

Comment:
match on `IntExpr` in the above function instead?

Reply:
Yup good call, originally I was going to support UInt64 but realised that doesn't make much sense